File: backend/cache.js
// Distributed cache over N Redis nodes, routed by consistent hashing.
// Cache-aside reads, write-around writes, jittered-TTL invalidation.
const Redis = require('ioredis');
const config = require('./config');
const metrics = require('./metrics');
const { ConsistentHashRing } = require('./consistentHash');

const clients = new Map();          // node -> ioredis client
let ring = new ConsistentHashRing(config.vnodes);

const redisKey = (prefix, mode) => `sugg:${mode}:${prefix}`;

async function init() {
  ring = new ConsistentHashRing(config.vnodes);
  for (const node of config.cacheNodes) {
    const [host, port] = node.split(':');
    clients.set(node, new Redis({ host, port: parseInt(port, 10), lazyConnect: false }));
    ring.addNode(node);
  }
  // fail fast if a node is unreachable
  for (const c of clients.values()) await c.ping();
}

async function close() {
  for (const c of clients.values()) c.disconnect();
}

function jittered(ttl) {
  const delta = ttl * config.ttlJitter;
  return Math.max(1, Math.round(ttl + (Math.random() * 2 - 1) * delta));
}

async function getSuggestions(prefix, mode) {
  const node = ring.getNode(prefix);
  const raw = await clients.get(node).get(redisKey(prefix, mode));
  if (raw == null) {
    metrics.m.cacheMisses++;
    return { suggestions: null, node, hit: false };
  }
  metrics.m.cacheHits++;
  return { suggestions: JSON.parse(raw), node, hit: true };
}

async function setSuggestions(prefix, mode, suggestions, ttl) {
  const node = ring.getNode(prefix);
  await clients.get(node).set(redisKey(prefix, mode), JSON.stringify(suggestions), 'EX', jittered(ttl));
  return node;
}

async function getRaw(key) {
  return clients.get(ring.getNode(key)).get(key);
}

async function setRaw(key, value, ttl) {
  const node = ring.getNode(key);
  await clients.get(node).set(key, value, 'EX', jittered(ttl));
  return node;
}

async function debug(prefix, mode) {
  const info = ring.debug(prefix);
  const node = info.owner_node;
  const present = node ? (await clients.get(node).get(redisKey(prefix, mode))) != null : false;
  return { ...info, mode, redis_key: redisKey(prefix, mode), currently_cached: present, hit_or_miss: present ? 'HIT' : 'MISS' };
}

module.exports = { init, close, getSuggestions, setSuggestions, getRaw, setRaw, debug, getRing: () => ring };

File: backend/cache.py
"""Distributed cache over N Redis nodes, routed by consistent hashing.
Cache-aside reads, write-around writes, jittered-TTL invalidation. The routing
key is the prefix, so both ranking modes for a prefix share one node."""
import json
import random
from typing import Dict, List, Optional, Tuple

import redis.asyncio as aioredis

from . import config, metrics
from .consistent_hash import ConsistentHashRing

_clients: Dict[str, "aioredis.Redis"] = {}
ring: ConsistentHashRing = ConsistentHashRing(vnodes=config.VNODES)


def _redis_key(prefix: str, mode: str) -> str:
    return f"sugg:{mode}:{prefix}"


async def init() -> None:
    global ring
    ring = ConsistentHashRing(vnodes=config.VNODES)
    for node in config.CACHE_NODES:
        host, port = node.split(":")
        _clients[node] = aioredis.Redis(host=host, port=int(port), decode_responses=True)
        ring.add_node(node)
    # fail fast if a node is unreachable
    for node, client in _clients.items():
        await client.ping()


async def close() -> None:
    for client in _clients.values():
        await client.aclose()


def _jittered(ttl: int) -> int:
    delta = ttl * config.TTL_JITTER
    return max(1, int(ttl + random.uniform(-delta, delta)))


async def get_suggestions(prefix: str, mode: str) -> Tuple[Optional[List], str, bool]:
    """Returns (suggestions|None, owner_node, hit)."""
    node = ring.get_node(prefix)
    client = _clients[node]
    raw = await client.get(_redis_key(prefix, mode))
    if raw is None:
        metrics.cache_misses += 1
        return None, node, False
    metrics.cache_hits += 1
    return json.loads(raw), node, True


async def set_suggestions(prefix: str, mode: str, suggestions: List, ttl: int) -> str:
    node = ring.get_node(prefix)
    client = _clients[node]
    await client.set(_redis_key(prefix, mode), json.dumps(suggestions), ex=_jittered(ttl))
    return node


async def get_raw(key: str) -> Optional[str]:
    node = ring.get_node(key)
    return await _clients[node].get(key)


async def set_raw(key: str, value: str, ttl: int) -> str:
    node = ring.get_node(key)
    await _clients[node].set(key, value, ex=_jittered(ttl))
    return node


async def debug(prefix: str, mode: str) -> dict:
    """Powers GET /cache/debug — routing + live hit/miss for the prefix key."""
    info = ring.debug(prefix)
    node = info["owner_node"]
    present = False
    if node is not None:
        present = (await _clients[node].get(_redis_key(prefix, mode))) is not None
    info["mode"] = mode
    info["redis_key"] = _redis_key(prefix, mode)
    info["currently_cached"] = present
    info["hit_or_miss"] = "HIT" if present else "MISS"
    return info
